File: basic_dsa_three_hundred_forty_three.java
/**
 * Basic DSA Problems - Integer Break (LeetCode 343)
 * Break n into sum of k>=2 positive integers, maximize product
 */
public class basic_dsa_three_hundred_forty_three {

    // Problem: n >= 2. Break into k>=2 positive integers. Maximize product.

    /**
     * DP: dp[i] = max product from breaking i. Try splitting at j: j * (i-j) or j * dp[i-j].
     * Time O(n^2), Space O(n)
     */
    public static int integerBreak(int n) {
        int[] dp = new int[n + 1];
        for (int i = 2; i <= n; i++) {
            for (int j = 1; j < i; j++)
                dp[i] = Math.max(dp[i], Math.max(j * (i - j), j * dp[i - j]));
        }
        return dp[n];
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Integer Break");
        System.out.println("=".repeat(60));

        System.out.println("2 -> " + integerBreak(2));
        System.out.println("10 -> " + integerBreak(10));
    }
}

File: basic_dsa_three_hundred_forty_three.cpp
/**
 * Basic DSA - Climbing Stairs (LeetCode 70)
 * Distinct ways to reach step n (1 or 2 steps at a time).
 */
#include <iostream>
#include <string>

int climbStairs(int n) {
    if (n <= 2) return n;
    int a = 1, b = 2;
    for (int i = 3; i <= n; i++) {
        int c = a + b;
        a = b;
        b = c;
    }
    return b;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Climbing Stairs\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "2 -> " << climbStairs(2) << "\n";
    std::cout << "3 -> " << climbStairs(3) << "\n";
    std::cout << "5 -> " << climbStairs(5) << "\n";
}
